src/core/components/files-regular/add-async-iterator.js

```diff
@@ -0,0 +1,200 @@
+'use strict'
+
+const importer = require('ipfs-unixfs-importer')
+const isSource = require('is-pull-stream').isSource
+const validateAddInput = require('ipfs-utils/src/files/add-input-validation')
+const { parseChunkerString } = require('./utils')
+const pipe = require('it-pipe')
+const { supportsFileReader } = require('ipfs-utils/src/supports')
+const isStream = require('is-stream')
+const toAsyncIterator = require('pull-stream-to-async-iterator')
+const log = require('debug')('ipfs:add')
+log.error = require('debug')('ipfs:add:error')
+
+function noop () {}
+
+module.exports = function (self) {
+  // Internal add func that gets used by all add funcs
+  return async function * addAsyncIterator (source, options) {
+    options = options || {}
+
+    let chunkerOptions = parseChunkerString(options.chunker)
+
+    const opts = Object.assign({}, {
+      shardSplitThreshold: self._options.EXPERIMENTAL.sharding
+        ? 1000
+        : Infinity
+    }, options, {
+      chunker: chunkerOptions.chunker,
+      chunkerOptions: chunkerOptions.chunkerOptions
+    })
+
+    // CID v0 is for multihashes encoded with sha2-256
+    if (opts.hashAlg && opts.cidVersion !== 1) {
+      opts.cidVersion = 1
+    }
+
+    let total = 0
+
+    const prog = opts.progress || noop
+    const progress = (bytes) => {
+      total += bytes
+      prog(total)
+    }
+
+    opts.progress = progress
+
+    if (Buffer.isBuffer(source) || typeof source === 'string') {
+      source = [
+        source
+      ]
+    }
+
+    const iterator = pipe(
+      source,
+      validateInput(),
+      normalizeInput(opts),
+      doImport(self, opts),
+      prepareFile(self, opts),
+      preloadFile(self, opts),
+      pinFile(self, opts)
+      // sort?
```

Comment:
We used to buffer & sort all the output.  This can lead to OOM with really, really large imports (think: package manager scale datasets) and `go-ipfs` doesn't do it, so 🤷‍♂